### DISCUSSION TITLE
Help Compiling Firmware

### DISCUSSION THREAD

**Discussion by feedbackl00p** [shanteacontrols/OpenDeck#208] [Q&A]
Im enjoying the hell out of tinkering with Opendeck, and Im trying to compile custom firmware for a Pico and getting errors. 

Im competent with electronics and MCUs, but not very competent with Linux or compiling. 

I think this is the process: 
- create custom yaml
- copy to config/target
- change to src directory
- run make TARGET=my-custom-conf

When I run this I get: 
![image](https://github.com/shanteacontrols/OpenDeck/assets/88065256/a6bfd326-10ac-4f82-9e59-40d8cb2b29bc)

I can compile the production firmware by going to /Opendeck and running 
make TARGET=pico. So I assume the components for make are OK. 

Any help is much appreciated. 

Edit: Ive also installed the opendeck Docker environment,  but Im not sure how to actually get into it.

**Comment by paradajz** [comment]
The info in the wiki was outdated, it's fixed now. You need to run the make command from the root directory of the repository, not from `src` anymore.

**Reply by feedbackl00p** [reply]
Perfect, thank you its working fine now. 

Im now struggling with the pin assignments for a 74hc595. Ive been through the repository and cant find a schematic where the 595 is implemented for a Pico. Im having issues reassigning pins and Id like to see a known working config for opendeck. Would you happen to have something like this? 

Also is there an explanation of the port types somewhere? Ive been through the wiki and cant work out what is the difference between an A and C port type and what kind of pin connection it requires. Im also struggling with the index number, is this always the pin/gpio number on the mcu?

Otherwise Ill just keep tinkering in this incredible playground youve created, thank you again for the reply.

**Comment by paradajz** [comment]
> Im now struggling with the pin assignments for a 74hc595. Ive been through the repository and cant find a schematic where the 595 is implemented for a Pico. Im having issues reassigning pins and Id like to see a known working config for opendeck. Would you happen to have something like this?

There is a config using 595 here:
https://github.com/shanteacontrols/OpenDeck/blob/master/config/target/dubfocus12.yml

What exactly are you struggling with?

> Also is there an explanation of the port types somewhere? Ive been through the wiki and cant work out what is the difference between an A and C port type and what kind of pin connection it requires.

There is none as far as the OpenDeck firmware is concerned.

> Im also struggling with the index number, is this always the pin/gpio number on the mcu?

All MCUs have GPIOs organized in a set of ports. Each port has several pins or indexes. For example on most AVR MCUs there is a PORTD which has pins 0-7. On STM32F4 MCUs ports usually have 16 pins available etc. It depends on the MCU. Again, I am not sure what are you struggling with?

**Reply by feedbackl00p** [reply]
Hey thanks for that, iv e already been right through the repository and looked at all the yamls and schematics but im trying to find a simple  implementation of the 595 specifically for the pico with the code and the mcu pin connections defined. 

The only ones I can find all use multi mux and dont have the register directly connected to mcu.

My issue is where do you find where the ports for a Pi Pico are defined?  ie are all ports on the Pico "0"?  Ive been through the datasheets and i can see is that there are two UART buses, but cant find how to define them in the opendeck doco. 

Im using this currently, but unable to get an attached led to light up when set to the same note as a button. Weirdly if i set the startup animation, the led does light during startup so it seems its attached, but the register isnt firing when pushing the button. Everything works normally when connecting as native components. 

Im using the same config below (with enable connected to ground, and pin 10 to vcc) with a micropython sketch, and the register is working correctly so i know the test rig is ok. Can you see anything obviously wrong? 

---

  mcu: "rp2040"

  usb: true

  buttons:

    type: "native"

    extPullups: false

    pins:

    -

      port: "0"

      index: 2

    leds:

    external:

      type: "shiftRegister"

      shiftRegisters: 1

      invert: false

      pins:

        data:

          port: "0"

          index: 9

        clock:

          port: "0"

          index: 13

        latch:

          port: "0"

          index: 8

        enable:

          port: "0"

          index: 12

**Comment by paradajz** [comment]
> Hey thanks for that, iv e already been right through the repository and looked at all the yamls and schematics but im trying to find a simple implementation of the 595 specifically for the pico with the code and the mcu pin connections defined.

Here you go:

![image](https://github.com/shanteacontrols/OpenDeck/assets/2544094/27596fc9-0794-421e-95ae-0bd6735620ad)

Connect data, clock, latch and enable on any pins you want.

> My issue is where do you find where the ports for a Pi Pico are defined?

In the datasheet?
https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf

Page 12 specifically.

> i can see is that there are two UART buses, but cant find how to define them in the opendeck doco.

https://github.com/shanteacontrols/OpenDeck/wiki/Raspberry-Pi

**Reply by feedbackl00p** [reply]
Thanks for that schematic, im actually looking for the pin conncections on the Pico so I can try to reverse engineer the code to understand the port and index.  Ive got the 595 working in python, I just cant get it working in Opendeck. Im trying to find a known working opendeck config with code and pin numbers. 

Im confident with building and coding basic microcontroller things and defining pins, but  Im still not understanding how you know what port definitions Opendeck is expecting in the code for a GPIO. ie the process of how to derive the opendeck port names and indexes from the pinouts and busses is eluding me. How to find out how Opendeck wants these defined is where Im stuck.

I can see there are two UART buses on the pico. ie GPIO 0 is on UART0 so is that port defined as '0' with index 0, and GPIO4 is on UART1 so that is port '1' and index 4. Is this correct? 

How then do you define non UART GPs? Do they have a different port number and where do I find this out? Is there a simple list of ports against GPIOs or is it contextual? 

Basically Im just trying to work out how to define the ports in opendeck with regards to the different busses and after several days Im not getting any closer. 

I feel like im missing something obvious or being very dense. 

This support is much appreciated mate, thank you.

**Comment by paradajz** [comment]
> Thanks for that schematic, im actually looking for the pin conncections on the Pico so I can try to reverse engineer the code to understand the port and index. 

Ehm, what? As I've mentioned in the post where I've posted the schematic, you can connect clock, latch, enable and data pins anywhere you want on Pico. You want to reverse engineer what exactly?

> Im still not understanding how you know what port definitions Opendeck is expecting in the code for a GPIO.

I seriously don't understand what do you want to know. This thread started with compiling issues, then moved on to 595 connection diagram and now it's low-level details such as this - I can't follow to be honest. Anyways, on RP2040 there is only one port, therefore it's named "0" in the config. Actually, since there is only one port, the port name doesn't even matter since I ignore it. You could name it "yaddayadda" and it will work.

For RP2040:
https://github.com/paradajz/core/blob/master/include/core/arch/arm/rpf/common/IO.h#L102

For STM32F4:
https://github.com/paradajz/core/blob/master/include/core/arch/arm/st/common/IO.h#L105

For nRF52:
https://github.com/paradajz/core/blob/master/include/core/arch/arm/nordic/common/IO.h#L100

etc.

> I can see there are two UART buses on the pico. ie GPIO 0 is on UART0 so is that port defined as '0' with index 0, and GPIO4 is on UART1 so that is port '1' and index 4. Is this correct?

No. Again, there is only one GPIO port on the RP2040. Therefore, GPIOx is always port "0" and index x.

> How then do you define non UART GPs? Do they have a different port number and where do I find this out? Is there a simple list of ports against GPIOs or is it contextual?

I have linked the wiki page where you can see which pins are allowed for UART. The default pinouts for supported boards are also listed in the wiki.

https://github.com/shanteacontrols/OpenDeck/wiki/Raspberry-Pi-Pico

**Reply by feedbackl00p** [reply]
Thanks for your help mate, its much appreciated.
